File: recyclerview/src/main/java/com/tunjid/androidbootstrap/recyclerview/diff/Differentiable.java
package com.tunjid.androidbootstrap.recyclerview.diff;


import com.tunjid.androidbootstrap.functions.Supplier;

public interface Differentiable {

    String getId();

    default boolean areContentsTheSame(Differentiable other) { return getId().equals(other.getId()); }

    default Object getChangePayload(Differentiable other) {
        return null;
    }

    static Differentiable fromCharSequence(Supplier<CharSequence> charSequenceSupplier) {
        final String id = charSequenceSupplier.get().toString();

        //noinspection EqualsWhichDoesntCheckParameterClass
        return new Differentiable() {
            @Override
            public boolean equals(Object obj) { return id.equals(obj); }

            @Override
            public int hashCode() { return id.hashCode(); }

            @Override
            public String getId() { return id; }
        };
    }

}

File: recyclerview/src/main/java/com/tunjid/androidbootstrap/recyclerview/diff/Differentiable.kt
package com.tunjid.androidbootstrap.recyclerview.diff


interface Differentiable {

    val diffId: String

    fun areContentsTheSame(other: Differentiable): Boolean = diffId == other.diffId

    fun getChangePayload(other: Differentiable): Any? = null

    companion object {

        fun fromCharSequence(charSequenceSupplier: () -> CharSequence): Differentiable {
            val id = charSequenceSupplier().toString()

            return object : Differentiable {
                override val diffId: String get() = id

                override fun equals(other: Any?): Boolean = id == other

                override fun hashCode(): Int = id.hashCode()
            }
        }
    }

}
